hazelcast/src/main/java/com/hazelcast/query/impl/BaseIndexStore.java

```diff
@@ -239,25 +266,177 @@ private Comparable sanitizeScalar(Object input) {
 
     }
 
-    private static class PassThroughFunctor implements CopyFunctor<Data, QueryableEntry> {
+    private class PassThroughFunctor implements CopyFunctor<Data, QueryableEntry> {
 
         @Override
         public Map<Data, QueryableEntry> invoke(Map<Data, QueryableEntry> map) {
+            if (isExpirable()) {
+                return new EvictionAwareHashMapDelegate(map);
+            }
             return map;
         }
 
     }
 
-    private static class CopyInputFunctor implements CopyFunctor<Data, QueryableEntry> {
+    private class CopyInputFunctor implements CopyFunctor<Data, QueryableEntry> {
 
         @Override
         public Map<Data, QueryableEntry> invoke(Map<Data, QueryableEntry> map) {
             if (map != null && !map.isEmpty()) {
-                return new HashMap<Data, QueryableEntry>(map);
+                HashMap<Data, QueryableEntry> newMap = new HashMap<Data, QueryableEntry>(map);
+                if (isExpirable()) {
+                    return new EvictionAwareHashMapDelegate(newMap);
+                }
+                return newMap;
             }
             return map;
         }
 
     }
 
+    /**
+     * This delegating Map updates the {@link Record}'s access time on every
+     * {@link QueryableEntry} retrieved through the {@link HashMap#entrySet()} or {@link HashMap#values()}.
+     *
+     */
+    private static class EvictionAwareHashMapDelegate implements Map<Data, QueryableEntry> {
```

Comment:
I think using Expiration instead of Eviction allover this PR fits better. Reasoning was same with https://github.com/hazelcast/hazelcast-enterprise/pull/3022#discussion_r293347727